# NOTEBOOK IMPORT AND PACKAGES

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## LOADING AND DESCRIBING THE DATA




In [7]:
# LOADING THE DATA

fuel_data= pd.read_csv('fuel_ferc1.csv', error_bad_lines= False)    

fuel_data.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

In [10]:
# DESCRIBING THE DATA
# this gives us the summary statistics like the mean, median, mode, standard deviation and percentiles(25th, 50th, 75th) etc


fuel_data.describe()  # this excludes all the Nan values

utility_id_ferc1   report_year  fuel_qty_burned  fuel_mmbtu_per_unit  \
count      29523.000000  29523.000000     2.952300e+04         29523.000000   
mean         118.601836   2005.806050     2.622119e+06             8.492111   
std           74.178353      7.025483     9.118004e+06            10.600220   
min            1.000000   1994.000000     1.000000e+00             0.000001   
25%           55.000000   2000.000000     1.381700e+04             1.024000   
50%          122.000000   2006.000000     2.533220e+05             5.762694   
75%          176.000000   2012.000000     1.424034e+06            17.006000   
max          514.000000   2018.000000     5.558942e+08           341.260000   

       fuel_cost_per_unit_burned  fuel_cost_per_unit_delivered  \
count               29523.000000                  2.952300e+04   
mean                  208.649031                  9.175704e+02   
std                  2854.490090                  6.877593e+04   
min                  -276.080000                 -8.749370e+02   
25%                     5.207000                  3.778500e+00   
50%                    26.000000                  1.737100e+01   
75%                    47.113000                  4.213700e+01   
max                139358.000000                  7.964521e+06   

       fuel_cost_per_mmbtu  
count         29523.000000  
mean             19.304354  
std            2091.540939  
min             -41.501000  
25%               1.940000  
50%               4.127000  
75%               7.745000  
max          359278.000000

# Question 1

If you are given two lists: 

A = [1, 2, 3, 4, 5, 6], B = [13, 21, 34] 

The task is to create a list with the dimesion of A and B in a single dimenstion with output: 

A_B = [1, 2, 3, 4, 5, 6, 13, 21, 34] 

Answer : None of the above

Explanation of my answer: the append and extend function only updates the value in an already existing list. They cannot be used to create a new list with a single dimension

# Question 2


How do you create an identity matrix

Answer: np.eye(3) 

An identity matrix is a matrix that has ones in its diagonal and every other value remains zero

Explanation of my answer: There were three possible answers in the option, but I chose the numpy eye function  because it can take different dimensions unlike the numpy identity function which only takes a value. Moreso, there was also a numpy array function which was also a possible answer

# Quesion 3

Which of the following fuel type code has the lowest average fuel cost per unit burned?

In [66]:
fuel_data.groupby('fuel_type_code_pudl')

lowest_avg_fuel_type= fuel_data.groupby('fuel_type_code_pudl')['fuel_cost_per_unit_burned'].mean().sort_values(ascending= False)
print(lowest_avg_fuel_type)


print('the fuel type code with the lowest average fuel cost per unit is Gas with value:', lowest_avg_fuel_type[5])

fuel_type_code_pudl
nuclear    4955.157002
oil         168.877086
coal         67.421830
waste        19.518122
other        18.253856
gas          13.659397
Name: fuel_cost_per_unit_burned, dtype: float64
the fuel type code with the lowest average fuel cost per unit is Gas with value: 13.659396569737135


Answer: Gas

# Question 4

What is the standard deviation and the 75th percentile of the measure of energy per unit (Fuel_mmbtu_per_unit) in two decimal places?

In [69]:
# standard deviation
std_fuel_mmbtu= fuel_data['fuel_mmbtu_per_unit'].std()
print('the standard deviation of fuel mmbtu per unit to 2dp:', np.around(std_fuel_mmbtu, 2))


# 75th percentile
fuel_mmbtu= np.array(fuel_data['fuel_mmbtu_per_unit'])
percentile_fuel_mmbtu= np.percentile(fuel_mmbtu, 75)

print('the 75th percentile of fuel mmbtu per unit to 2dp:', np.around(percentile_fuel_mmbtu, 2))



the standard deviation of fuel mmbtu per unit to 2dp: 10.6
the 75th percentile of fuel mmbtu per unit to 2dp: 17.01


Answer: 10.6 and 17.01

# Question 5

What is the skewness and kurtosis for the fuel quality burned in two decimal places?

In [68]:
#skewness
fuel_qty_burned= fuel_data['fuel_qty_burned']

skew_fuel_qty_burned= fuel_qty_burned.skew()

# kurtosis
kurtosis_fuel_qty_burned= fuel_qty_burned.kurtosis()

#Result
print('the skewness of fuel qty burned to 2dp:', np.around(skew_fuel_qty_burned, 2) )
print('the kurtosis of fuel qty burned to 2dp:', np.around(kurtosis_fuel_qty_burned, 2))

the skewness of fuel qty burned to 2dp: 15.85
the kurtosis of fuel qty burned to 2dp: 651.37


Answer: 15.85 and 651.37

# Question 6
which of the features has missing values and what is the total number of missing value and percentage of the misssing rows as a factor of the total number of rows in three decimal places?

In [70]:
# the feature
feature_missing= fuel_data.isnull().any().index[5]

print('the feature with missing values:', feature_missing)

# the total number of missing values
num_missing_vals= fuel_data.isnull().sum()[5]

print('the total number of missing values', num_missing_vals)

# the percentage of the missing rows as a factor of the total number of rows
percent= (num_missing_vals / fuel_data.shape[0]) * 100

print('the percentage of the missing row',percent)

the feature with missing values: fuel_unit
the total number of missing values 180
the percentage of the missing row 0.609694136774718


In [45]:
df= pd.DataFrame({'Feature': feature_missing, 'Total': num_missing_vals, 'Percent': round(percent,3)}, index= [0])

print(df)

     Feature  Total  Percent
0  fuel_unit    180     0.61


Answer: 
    
Feature: fuel_unit, Total: 180, Percent: 0.61

# Question 7

The feature with missing values falls under what category? What missing value imputation technique would you use?

In [49]:
fuel_data.isnull().any()

record_id                       False
utility_id_ferc1                False
report_year                     False
plant_name_ferc1                False
fuel_type_code_pudl             False
fuel_unit                        True
fuel_qty_burned                 False
fuel_mmbtu_per_unit             False
fuel_cost_per_unit_burned       False
fuel_cost_per_unit_delivered    False
fuel_cost_per_mmbtu             False
dtype: bool

In [53]:
fuel_data['fuel_unit'].isnull().any() # fuel_unit isnull value is true which shows that there are missing values in the column

True

Answer: None of the above
    
Explanation of my answer: 

from the above, the feature with missing value is fuel_unit which falls under the categorical data. 

the missing value imputation will be: fuel_data['fuel_unit'] = fuel_data['fuel_unit'].fillna('a string or value')


# Question 8

which of the features has the second and third lowest correlation with the fuel cost per unit burned?

In [81]:
fuel_data.corr().loc['fuel_cost_per_unit_burned'].sort_values(ascending= False)

fuel_cost_per_unit_burned       1.000000
report_year                     0.013599
fuel_cost_per_unit_delivered    0.011007
fuel_cost_per_mmbtu            -0.000437
fuel_mmbtu_per_unit            -0.010034
fuel_qty_burned                -0.018535
utility_id_ferc1               -0.037863
Name: fuel_cost_per_unit_burned, dtype: float64

Answer: fuel_qty_burned and fuel_mmbtu_per_unit

# Question 9

For the fuel type coal, what is the percentage change in the fuel cost per unit burned in 1998 compared to 1994?

In [83]:
cost_years= fuel_data.groupby(['report_year', 'fuel_type_code_pudl'])['fuel_cost_per_unit_burned'].sum()

print(cost_years)

report_year  fuel_type_code_pudl
1994         coal                    14984.572
             gas                     10792.496
             nuclear                227983.354
             oil                      9362.194
             other                     116.737
                                       ...    
2018         coal                     9488.197
             gas                      4322.494
             nuclear                173267.161
             oil                     33836.602
             waste                     660.972
Name: fuel_cost_per_unit_burned, Length: 148, dtype: float64


In [85]:
#coal 1994
coal_year_1994= cost_years.loc[1994, 'coal']

print('fuel type coal 1994:', coal_year_1994)

# coal 1998
coal_year_1998= cost_years.loc[1998, 'coal']

print('fuel type coal 1998:', coal_year_1994)

# percentage change

percentage_change= ((coal_year_1998 - coal_year_1994) / coal_year_1994) * 100

print('the percentage change in the fuel cost per unit burned:', percentage_change)


fuel type coal 1994: 14984.572000000011
fuel type coal 1998: 14984.572000000011
the percentage change in the fuel cost per unit burned: -20.56765451826049


Answer: -20.56765451826049

# Question 10

what year has the highest average fuel cost per unit delivered?

In [63]:
fuel_data.groupby('report_year')

high_avg_fuel_cost= fuel_data.groupby('report_year')['fuel_cost_per_unit_delivered'].mean().sort_values(ascending= False)

print(high_avg_fuel_cost)
print('the year with the highest average fuel cost per unit delivered is 1997 with value: ', high_avg_fuel_cost.iloc[0])

report_year
1997    11140.197239
1996     9196.705948
2000      985.362877
2009      652.694163
2018      499.269966
2015      326.535511
1998      287.154420
2014      192.737183
2013      172.307591
2004      139.524275
2016      103.901761
2010       91.862105
1994       63.636060
2012       60.994502
2001       60.050396
2011       59.774667
2008       58.588197
2003       55.663493
2002       47.594361
2017       46.196861
2007       43.325023
2005       41.438184
2006       38.657484
1995       32.735269
1999       25.551627
Name: fuel_cost_per_unit_delivered, dtype: float64
the year with the highest average fuel cost per unit delivered is 1997 with value:  11140.19723948813


Answer: 1997